File: server/src/retrieval/sourceAttribution.js
// Real implementation — extends sourceFormatter.js's filename-only citation
// with page/line ranges and a confidence score derived from similarity, per
// the target response shape in ARCHITECTURE.md section 6.
export function attributeSources(chunks) {
  const seen = new Set();
  const sources = [];

  for (const chunk of chunks) {
    const filename = chunk.original_filename || 'unknown source';
    const key = `${filename}:${chunk.page_number ?? ''}:${chunk.line_start ?? ''}-${chunk.line_end ?? ''}`;
    if (seen.has(key)) continue;
    seen.add(key);

    const source = {
      index: sources.length + 1,
      filename,
      type: chunk.source_type || 'document',
      confidence: chunk.similarity != null ? Number(chunk.similarity.toFixed(2)) : null
    };
    if (chunk.page_number != null) source.page = chunk.page_number;
    if (chunk.line_start != null && chunk.line_end != null) {
      source.lines = `${chunk.line_start}-${chunk.line_end}`;
    }
    sources.push(source);
  }

  return sources;
}

File: server/app/retrieval/source_attribution.py
def attribute_sources(chunks: list[dict]) -> list[dict]:
    seen = set()
    sources = []

    for chunk in chunks:
        filename = chunk.get("original_filename") or "unknown source"
        key = f"{filename}:{chunk.get('page_number', '')}:{chunk.get('line_start', '')}-{chunk.get('line_end', '')}"
        if key in seen:
            continue
        seen.add(key)

        similarity = chunk.get("similarity")
        source = {
            "index": len(sources) + 1,
            "chunk_id": chunk.get("id"),
            "filename": filename,
            "type": chunk.get("source_type") or "document",
            "similarity": round(similarity, 4) if similarity is not None else None,
            "confidence": round(similarity, 2) if similarity is not None else None,
        }
        if chunk.get("page_number") is not None:
            source["page"] = chunk["page_number"]
        if chunk.get("line_start") is not None and chunk.get("line_end") is not None:
            source["lines"] = f"{chunk['line_start']}-{chunk['line_end']}"
        sources.append(source)

    return sources
